File: frontend/test_presets.js
/**
 * Script de test pour vérifier la génération des dates des presets côté frontend
 * Exécuter avec: node test_presets.js
 */

function generateDatesFromPresets(presets) {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  
  const dates_depart = [];
  const dates_retour = [];
  const addedDatesDepart = new Set();
  const addedDatesRetour = new Set();
  
  const addDateIfNotExists = (date, isDepart) => {
    if (isDepart) {
      if (!addedDatesDepart.has(date.date)) {
        addedDatesDepart.add(date.date);
        dates_depart.push(date);
      }
    } else {
      if (!addedDatesRetour.has(date.date)) {
        addedDatesRetour.add(date.date);
        dates_retour.push(date);
      }
    }
  };
  
  const validPresets = presets.filter(p => p !== 'flexible');
  
  validPresets.forEach(preset => {
    if (preset === 'weekend') {
      // Ce weekend : vendredi-dimanche prochain
      const dayOfWeek = today.getDay(); // 0 = dimanche, 5 = vendredi
      let daysUntilFriday = (5 - dayOfWeek + 7) % 7;
      if (daysUntilFriday === 0) {
        daysUntilFriday = 7; // Si on est déjà vendredi, prendre le suivant
      }
      
      const friday = new Date(today.getTime() + daysUntilFriday * 24 * 60 * 60 * 1000);
      const sunday = new Date(friday.getTime() + 2 * 24 * 60 * 60 * 1000);
      
      addDateIfNotExists({
        date: friday.toISOString().split('T')[0],
        heure_min: '06:00',
        heure_max: '23:59'
      }, true);
      addDateIfNotExists({
        date: sunday.toISOString().split('T')[0],
        heure_min: '06:00',
        heure_max: '23:59'
      }, false);
    } else if (preset === 'next-weekend') {
      // Weekend prochain : vendredi-dimanche suivant
      const dayOfWeek = today.getDay();
      let daysUntilFriday = (5 - dayOfWeek + 7) % 7;
      if (daysUntilFriday === 0) {
        daysUntilFriday = 7; // Si on est vendredi, prendre le suivant
      }
      daysUntilFriday += 7; // Toujours ajouter 7 pour le weekend prochain
      
      const friday = new Date(today.getTime() + daysUntilFriday * 24 * 60 * 60 * 1000);
      const sunday = new Date(friday.getTime() + 2 * 24 * 60 * 60 * 1000);
      
      addDateIfNotExists({
        date: friday.toISOString().split('T')[0],
        heure_min: '06:00',
        heure_max: '23:59'
      }, true);
      addDateIfNotExists({
        date: sunday.toISOString().split('T')[0],
        heure_min: '06:00',
        heure_max: '23:59'
      }, false);
    } else if (preset === 'next-week') {
      // 3 jours la semaine prochaine : vendredi-dimanche
      // Horaires : 23h00 à 6h00 (plage qui traverse minuit)
      const dayOfWeek = today.getDay(); // 0 = dimanche, 5 = vendredi
      let daysUntilFriday = (5 - dayOfWeek + 7) % 7;
      if (daysUntilFriday === 0) {
        daysUntilFriday = 7; // Si on est vendredi, prendre le suivant
      }
      daysUntilFriday += 7; // Toujours ajouter 7 pour la semaine prochaine
      
      const friday = new Date(today.getTime() + daysUntilFriday * 24 * 60 * 60 * 1000);
      const sunday = new Date(friday.getTime() + 2 * 24 * 60 * 60 * 1000);
      
      addDateIfNotExists({
        date: friday.toISOString().split('T')[0],
        heure_min: '23:00',
        heure_max: '06:00'
      }, true);
      addDateIfNotExists({
        date: sunday.toISOString().split('T')[0],
        heure_min: '23:00',
        heure_max: '06:00'
      }, false);
    }
  });
  
  // Trier les dates
  dates_depart.sort((a, b) => a.date.localeCompare(b.date));
  dates_retour.sort((a, b) => a.date.localeCompare(b.date));
  
  return { dates_depart, dates_retour };
}

function getDayName(dateStr) {
  const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  const date = new Date(dateStr + 'T00:00:00');
  return days[date.getDay()];
}

function testPreset(presetName) {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`Test du preset: ${presetName}`);
  console.log('='.repeat(60));
  
  try {
    const { dates_depart, dates_retour } = generateDatesFromPresets([presetName]);
    
    console.log(`\nDates de départ (${dates_depart.length}):`);
    dates_depart.forEach(d => {
      const dayName = getDayName(d.date);
      console.log(`  - ${d.date} (${dayName}) - ${d.heure_min} à ${d.heure_max}`);
    });
    
    console.log(`\nDates de retour (${dates_retour.length}):`);
    dates_retour.forEach(d => {
      const dayName = getDayName(d.date);
      console.log(`  - ${d.date} (${dayName}) - ${d.heure_min} à ${d.heure_max}`);
    });
    
    // Vérifications
    const errors = [];
    
    // Vérifier que tous les départs sont le vendredi
    dates_depart.forEach(d => {
      const dayName = getDayName(d.date);
      if (dayName !== 'Friday') {
        errors.push(`❌ ERREUR: Le départ ${d.date} n'est pas un vendredi (c'est un ${dayName})`);
      }
    });
    
    // Vérifier que tous les retours sont le dimanche
    dates_retour.forEach(d => {
      const dayName = getDayName(d.date);
      if (dayName !== 'Sunday') {
        errors.push(`❌ ERREUR: Le retour ${d.date} n'est pas un dimanche (c'est un ${dayName})`);
      }
    });
    
    // Vérifier qu'il y a au moins une date de départ et une de retour
    if (dates_depart.length === 0) {
      errors.push('❌ ERREUR: Aucune date de départ générée');
    }
    if (dates_retour.length === 0) {
      errors.push('❌ ERREUR: Aucune date de retour générée');
    }
    
    // Vérifier qu'il n'y a qu'un seul départ et un seul retour
    if (dates_depart.length > 1) {
      errors.push(`⚠️  ATTENTION: ${dates_depart.length} dates de départ générées (attendu: 1)`);
    }
    if (dates_retour.length > 1) {
      errors.push(`⚠️  ATTENTION: ${dates_retour.length} dates de retour générées (attendu: 1)`);
    }
    
    if (errors.length > 0) {
      console.log(`\n${'!'.repeat(60)}`);
      console.log('PROBLÈMES DÉTECTÉS:');
      errors.forEach(error => console.log(`  ${error}`));
      console.log('!'.repeat(60));
      return false;
    } else {
      console.log(`\n✅ Tous les tests sont passés pour ${presetName}!`);
      return true;
    }
  } catch (error) {
    console.log(`\n❌ ERREUR lors du test: ${error.message}`);
    console.error(error);
    return false;
  }
}

function main() {
  console.log('='.repeat(60));
  console.log('TEST DES PRESETS DE DATES (FRONTEND)');
  console.log('='.repeat(60));
  const today = new Date();
  console.log(`Date actuelle: ${today.toISOString().split('T')[0]} (${getDayName(today.toISOString().split('T')[0])})`);
  
  const presets = ['weekend', 'next-weekend', 'next-week'];
  
  const results = {};
  presets.forEach(preset => {
    results[preset] = testPreset(preset);
  });
  
  // Résumé
  console.log(`\n${'='.repeat(60)}`);
  console.log('RÉSUMÉ DES TESTS');
  console.log('='.repeat(60));
  Object.entries(results).forEach(([preset, success]) => {
    const status = success ? '✅ PASSÉ' : '❌ ÉCHOUÉ';
    console.log(`  ${preset}: ${status}`);
  });
  
  const allPassed = Object.values(results).every(r => r);
  if (allPassed) {
    console.log(`\n🎉 Tous les tests sont passés!`);
  } else {
    console.log(`\n⚠️  Certains tests ont échoué. Vérifiez les erreurs ci-dessus.`);
  }
  
  process.exit(allPassed ? 0 : 1);
}

main();

File: backend/test_presets.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Script de test pour vérifier la génération des dates des presets
"""
from datetime import date, timedelta
from typing import List, Tuple

class DateAvecHoraire:
    def __init__(self, date: str, heure_min: str, heure_max: str):
        self.date = date
        self.heure_min = heure_min
        self.heure_max = heure_max
    
    def __repr__(self):
        return f"DateAvecHoraire(date='{self.date}', heure_min='{self.heure_min}', heure_max='{self.heure_max}')"

def get_dates_from_preset(preset: str) -> Tuple[List[DateAvecHoraire], List[DateAvecHoraire]]:
    """
    Convertit un preset de dates en listes de DateAvecHoraire pour aller et retour
    """
    today = date.today()
    dates_depart = []
    dates_retour = []
    
    if preset == 'weekend':
        # Ce weekend : vendredi-dimanche prochain
        days_until_friday = (4 - today.weekday()) % 7
        if days_until_friday == 0 and today.weekday() >= 4:
            days_until_friday = 7  # Si on est déjà vendredi ou après, prendre le suivant
        
        friday = today + timedelta(days=days_until_friday)
        sunday = friday + timedelta(days=2)
        
        dates_depart.append(DateAvecHoraire(date=friday.isoformat(), heure_min="06:00", heure_max="23:59"))
        dates_retour.append(DateAvecHoraire(date=sunday.isoformat(), heure_min="06:00", heure_max="23:59"))
        
    elif preset == 'next-weekend':
        # Weekend prochain : vendredi-dimanche suivant
        days_until_friday = (4 - today.weekday()) % 7
        if days_until_friday == 0:
            days_until_friday = 7
        else:
            days_until_friday += 7
        
        friday = today + timedelta(days=days_until_friday)
        sunday = friday + timedelta(days=2)
        
        dates_depart.append(DateAvecHoraire(date=friday.isoformat(), heure_min="06:00", heure_max="23:59"))
        dates_retour.append(DateAvecHoraire(date=sunday.isoformat(), heure_min="06:00", heure_max="23:59"))
        
    elif preset == 'next-week':
        # 3 jours la semaine prochaine : vendredi-dimanche
        # Horaires : 23h00 à 6h00 (plage qui traverse minuit)
        days_until_friday = (4 - today.weekday()) % 7
        if days_until_friday == 0 and today.weekday() >= 4:
            days_until_friday = 7  # Si on est déjà vendredi ou après, prendre le suivant
        else:
            days_until_friday += 7  # Semaine prochaine
        
        friday = today + timedelta(days=days_until_friday)
        sunday = friday + timedelta(days=2)
        
        dates_depart.append(DateAvecHoraire(date=friday.isoformat(), heure_min="23:00", heure_max="06:00"))
        dates_retour.append(DateAvecHoraire(date=sunday.isoformat(), heure_min="23:00", heure_max="06:00"))
    
    return dates_depart, dates_retour

def test_preset(preset_name: str):
    """Teste un preset et affiche les resultats"""
    print(f"\n{'='*60}")
    print(f"Test du preset: {preset_name}")
    print(f"{'='*60}")
    
    try:
        dates_depart, dates_retour = get_dates_from_preset(preset_name)
        
        print(f"\nDates de depart ({len(dates_depart)}):")
        for d in dates_depart:
            date_obj = date.fromisoformat(d.date)
            day_name = date_obj.strftime("%A")
            print(f"  - {d.date} ({day_name}) - {d.heure_min} a {d.heure_max}")
        
        print(f"\nDates de retour ({len(dates_retour)}):")
        for d in dates_retour:
            date_obj = date.fromisoformat(d.date)
            day_name = date_obj.strftime("%A")
            print(f"  - {d.date} ({day_name}) - {d.heure_min} a {d.heure_max}")
        
        # Verifications
        errors = []
        
        # Verifier que tous les departs sont le vendredi
        for d in dates_depart:
            date_obj = date.fromisoformat(d.date)
            if date_obj.weekday() != 4:  # 4 = vendredi
                errors.append(f"ERREUR: Le depart {d.date} n'est pas un vendredi (c'est un {date_obj.strftime('%A')})")
        
        # Verifier que tous les retours sont le dimanche
        for d in dates_retour:
            date_obj = date.fromisoformat(d.date)
            if date_obj.weekday() != 6:  # 6 = dimanche
                errors.append(f"ERREUR: Le retour {d.date} n'est pas un dimanche (c'est un {date_obj.strftime('%A')})")
        
        # Verifier qu'il y a au moins une date de depart et une de retour
        if len(dates_depart) == 0:
            errors.append("ERREUR: Aucune date de depart generee")
        if len(dates_retour) == 0:
            errors.append("ERREUR: Aucune date de retour generee")
        
        # Verifier qu'il n'y a qu'un seul depart et un seul retour
        if len(dates_depart) > 1:
            errors.append(f"ATTENTION: {len(dates_depart)} dates de depart generees (attendu: 1)")
        if len(dates_retour) > 1:
            errors.append(f"ATTENTION: {len(dates_retour)} dates de retour generees (attendu: 1)")
        
        if errors:
            print(f"\n{'!'*60}")
            print("PROBLEMES DETECTES:")
            for error in errors:
                print(f"  {error}")
            print(f"{'!'*60}")
            return False
        else:
            print(f"\nTous les tests sont passes pour {preset_name}!")
            return True
        
    except Exception as e:
        print(f"\nERREUR lors du test: {e}")
        import traceback
        traceback.print_exc()
        return False

def main():
    """Fonction principale"""
    print("="*60)
    print("TEST DES PRESETS DE DATES")
    print("="*60)
    print(f"Date actuelle: {date.today()} ({date.today().strftime('%A')})")
    
    presets = ['weekend', 'next-weekend', 'next-week']
    
    results = {}
    for preset in presets:
        results[preset] = test_preset(preset)
    
    # Resume
    print(f"\n{'='*60}")
    print("RESUME DES TESTS")
    print(f"{'='*60}")
    for preset, success in results.items():
        status = "PASSE" if success else "ECHOUE"
        print(f"  {preset}: {status}")
    
    all_passed = all(results.values())
    if all_passed:
        print(f"\nTous les tests sont passes!")
    else:
        print(f"\nCertains tests ont echoue. Verifiez les erreurs ci-dessus.")
    
    return 0 if all_passed else 1

if __name__ == "__main__":
    exit(main())
